scilpy/tractograms/tractogram_operations.py

```diff
@@ -635,35 +649,46 @@ def upsample_tractogram(sft, nb, point_wise_std=None, streamline_wise_std=None,
     new_sft : StatefulTractogram
         The upsampled tractogram.
     """
-    assert bool(point_wise_std) ^ bool(streamline_wise_std), \
-        'Can only add either point-wise or streamline-wise noise' + \
-        ', not both nor none.'
-
-    rng = np.random.RandomState(seed)
-
-    # Get the number of streamlines to add
-    nb_new = nb - len(sft.streamlines)
+    rng = np.random.default_rng(seed)
 
     # Get the streamlines that will serve as a base for new ones
-    indices = rng.choice(
-        len(sft.streamlines), nb_new)
-    new_streamlines = sft.streamlines.copy()
+    resample_sft = resample_streamlines_step_size(sft, 1)
+    new_streamlines = []
+    indices = rng.choice(len(resample_sft), nb, replace=True)
+    unique_indices, count = np.unique(indices, return_counts=True)
 
     # For all selected streamlines, add noise and smooth
-    for s in sft.streamlines[indices]:
-        if point_wise_std:
-            noise = rng.normal(scale=point_wise_std, size=s.shape)
-        else:  # streamline_wise_std
-            noise = rng.normal(scale=streamline_wise_std, size=s.shape[-1])
-        new_s = s + noise
+    for i, c in zip(unique_indices, count):
+        s = resample_sft.streamlines[i]
+        if len(s) < 3:
+            new_streamlines.extend(np.repeat(s, c).tolist())
+        new_s = parallel_transport_streamline(s, c, tube_radius)
+
+        # Generate smooth noise_factor
+        noise = rng.normal(loc=0, scale=point_wise_std,
+                           size=len(s))
+
+        x = np.arange(len(noise))
+        poly_coeffs = np.polyfit(x, noise, 3)
+        polynomial = Polynomial(poly_coeffs[::-1])
```

Comment:
Can you please add some explanation of what's going on? Ex, the word parallel transport does not appear in this method. Can you explain how and why plynomials are used?

Thanks!